src/Periphery/ServiceFeeCollector.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.17;
+
+import { LibAsset } from "../Libraries/LibAsset.sol";
+import { TransferrableOwnership } from "../Helpers/TransferrableOwnership.sol";
+
+/// @title Service Fee Collector
+/// @author LI.FI (https://li.fi)
+/// @notice Provides functionality for collecting service fees (gas/insurance)
+contract ServiceFeeCollector is TransferrableOwnership {
+    /// Errors ///
+    error TransferFailure();
+    error NotEnoughNativeForFees();
+
+    /// Events ///
+    event GasFeesCollected(
+        address indexed _token,
+        address indexed _receiver,
+        uint256 _feeAmount
+    );
+
+    event InsuranceFeesCollected(
+        address indexed _token,
+        address indexed _receiver,
+        uint256 _feeAmount
+    );
+
+    event FeesWithdrawn(
+        address indexed _token,
+        address indexed _to,
+        uint256 _amount
+    );
+
+    /// Constructor ///
+
+    constructor(address _owner) TransferrableOwnership(_owner) {}
+
+    /// External Methods ///
+
+    /// @notice Collects gas fees
+    /// @param tokenAddress The address of the token to collect
+    /// @param feeAmount The amount of fees to collect
+    /// @param receiver The address to send gas to on the destination chain
+    function collectTokenGasFees(
+        address tokenAddress,
+        uint256 feeAmount,
+        address receiver
+    ) external {
+        LibAsset.depositAsset(tokenAddress, feeAmount);
+        emit GasFeesCollected(tokenAddress, receiver, feeAmount);
+    }
+
+    /// @notice Collects gas fees in native token
+    /// @param feeAmount The amount of native token to collect
+    /// @param receiver The address to send gas to on destination chain
+    function collectNativeGasFees(uint256 feeAmount, address receiver)
+        external
+        payable
+    {
+        if (msg.value < feeAmount) revert NotEnoughNativeForFees();
+        uint256 remaining = msg.value - (feeAmount);
+        // Prevent extra native token from being locked in the contract
+        if (remaining > 0) {
+            (bool success, ) = payable(msg.sender).call{ value: remaining }(
+                ""
+            );
+            if (!success) {
+                revert TransferFailure();
+            }
+        }
+        emit GasFeesCollected(LibAsset.NULL_ADDRESS, receiver, feeAmount);
```

Comment:
I imagine it would be cheaper to use address(0) instead of using Library code for this, but I might be wrong